x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/PushDownAndCombineLimits.java

```diff
@@ -62,16 +54,17 @@ public LogicalPlan rule(Limit limit, LogicalOptimizerContext ctx) {
                     var l1 = (int) limit.limit().fold(ctx.foldCtx());
                     var l2 = (int) descendantLimit.limit().fold(ctx.foldCtx());
                     if (l2 <= l1) {
-                        return new Limit(limit.source(), Literal.of(limit.limit(), l2), limit.child());
+                        return limit.withLimit(Literal.of(limit.limit(), l2));
                     }
                 }
             }
-        } else if (limit.child() instanceof Join join) {
+        } else if (limit.child() instanceof Join join && limit.duplicated() == false) {
             if (join.config().type() == JoinTypes.LEFT) {
-                // NOTE! This is only correct because our LEFT JOINs preserve the number of rows from the left hand side.
-                // This deviates from SQL semantics. In SQL, multiple matches on the right hand side lead to multiple rows in the output.
-                // For us, multiple matches on the right hand side are collected into multi-values.
-                return join.replaceChildren(limit.replaceChild(join.left()), join.right());
+                // Left joins increase the number of rows if any join key has multiple matches from the right hand side.
+                // Therefore, we cannot simply push down the limit - but we can add another limit before the join.
+                // To avoid repeating this infinitely, we have to set duplicated = true.
+                Join newChild = join.replaceChildren(limit.replaceChild(join.left()), join.right());
+                return limit.replaceChild(newChild).withDuplicated(true);
```

Comment:
Wondering if introducing a static method `duplicateLimit(Limit, Supplier<LogicalNode>)` would make the duplication more incapsulated.